vyper/builtin_functions/functions.py

```diff
@@ -1256,40 +1335,56 @@ def build_IR(self, expr, args, kwargs, contact):
 class RawLog(_BuiltinFunction):
 
     _id = "raw_log"
-    _inputs = [("topics", "*"), ("data", ("bytes32", "Bytes"))]
+    _inputs = [
+        ("topics", DynamicArrayDefinition(Bytes32Definition(), 4)),
+        ("data", (Bytes32Definition(), BytesArrayPrimitive())),
+    ]
 
     def fetch_call_return(self, node):
+        self.infer_arg_types(node)
+
+    def infer_arg_types(self, node):
         validate_call_args(node, 2)
         if not isinstance(node.args[0], vy_ast.List) or len(node.args[0].elements) > 4:
             raise InvalidType("Expecting a list of 0-4 topics as first argument", node.args[0])
+
         if node.args[0].elements:
-            validate_expected_type(
-                node.args[0], ArrayDefinition(Bytes32Definition(), len(node.args[0].elements))
+            validate_expected_type(node.args[0], self._inputs[0][1])
+
+        data_type = get_possible_types_from_node(node.args[1]).pop()
+
+        if not isinstance(data_type, BytesAbstractType):
+            expected_str = " or ".join(str(i) for i in self._inputs[1][1])
+            raise InvalidType(
+                f"Expected {expected_str} but value can only be cast as {data_type}", node.args[1]
             )
-        validate_expected_type(node.args[1], BytesAbstractType())
 
-    @validate_inputs
-    def build_IR(self, expr, args, kwargs, context):
-        if not isinstance(args[0], vy_ast.List) or len(args[0].elements) > 4:
-            raise StructureException("Expecting a list of 0-4 topics as first argument", args[0])
+        return [self._inputs[0][1], data_type]
+
+    def build_IR(self, expr, context):
         topics = []
-        for elt in args[0].elements:
-            arg = Expr.parse_value_expr(elt, context)
-            if not is_base_type(arg.typ, "bytes32"):
-                raise TypeMismatch("Expecting a bytes32 argument as topic", elt)
-            topics.append(arg)
-        if args[1].typ == BaseType("bytes32"):
+        for a in expr.args[0].elements:
```

Comment:
hm, now that `topics` is `DynamicArrayDefinition`, will `@validate_inputs` work on it again?

Reply:
I think it still does not work because the topics argument is a `List` AST node, and the expected output is a list of IR nodes whereas `process_arg` returns an IR node.